### PR TITLE
feat: expose all user-tunable env vars in the settings UI (add-on parity)

### LINKED ISSUE #1538: [FEATURE] Expose all user-tunable env vars in the advanced web settings UI (add-on parity)
## What would you like to see?

The advanced web settings UI should expose every *user-tunable* env var, so add-on users (who can't set raw environment variables) have parity with Docker/stdio/uv users who can. The settings UI was built for exactly this, but it only surfaces vars **registered** in `config.py` (`ADVANCED_SETTINGS_FIELDS` / `FEATURE_FLAG_FIELDS` / `BACKUP_OVERRIDE_FIELDS` + the `Config` model). Any env var read outside that registry is invisible to add-on users.

Concretely, these tunable knobs are read directly via `os.environ` in `tools/smart_search/_config.py`, bypass `config.py`, and so do **not** appear in the UI:

- `HAMCP_AUTOMATION_CONFIG_TIME_BUDGET` (default 30s)
- `HAMCP_SCRIPT_CONFIG_TIME_BUDGET` (default 20s)
- `HAMCP_SCENE_CONFIG_TIME_BUDGET` (default 20s)

Ask:

1. Register these three so they ride the existing UI plumbing.
2. Audit the full env-var surface and expose every remaining user-tunable one.
3. Add a guard (test) so a new `os.environ` read that isn't registered fails CI — prevents this drift recurring.

## Why do you need this?

Add-on users can't set env vars, so any unregistered one is permanently inaccessible to them — breaking the parity the settings UI was created to deliver. It just became user-visible: `ha_search` now returns `partial_reason: "… tune HAMCP_AUTOMATION_CONFIG_TIME_BUDGET to raise the budget"`, but add-on users have no way to act on that hint.

## Additional context

Audit — env vars read in `src/` but not registered in `config.py`:

- **Tunable behavior (should be exposed):** `HAMCP_AUTOMATION_CONFIG_TIME_BUDGET`, `HAMCP_SCRIPT_CONFIG_TIME_BUDGET`, `HAMCP_SCENE_CONFIG_TIME_BUDGET`
- **Review (likely env-only by design — bind/secret/bootstrap/path):** `FASTMCP_PORT`, `FASTMCP_TRANSPORT`, `MCP_HOST`, `MCP_PORT`, `MCP_HTTP_PORT`, `MCP_BASE_URL`, `MCP_SECRET_PATH`, `SUPERVISOR_BASE_URL`, `HA_MCP_CONFIG_DIR`, `HA_MCP_DISABLE_SETTINGS_UI`, `HA_MCP_BUILD_VERSION`, `XDG_DATA_HOME`

Note: the three budget vars are read at import as module-level constants, so exposing them via the UI should account for the existing "changes require an MCP-host restart" behavior.

_(Reported with analysis assisted by Claude.)_

### PR DESCRIPTION
**Pull request by kingpanther13**
## What does this PR do?

Closes #1538. The web Settings UI only surfaces env vars registered in
`config.py`, so add-on users (who can't set raw environment variables) couldn't
reach any tunable that lived outside that registry. This audits the full
env-var surface, exposes every remaining **user-tunable** one, and adds CI
guards so the gap can't reopen.

**Newly exposed in the Advanced panel:**

- `HAMCP_AUTOMATION_CONFIG_TIME_BUDGET`, `HAMCP_SCRIPT_CONFIG_TIME_BUDGET`,
  `HAMCP_SCENE_CONFIG_TIME_BUDGET` — read directly via `os.environ` in
  `tools/smart_search/_config.py`, bypassing `config.py` entirely. Now
  first-class `Settings` fields + `ADVANCED_SETTINGS_FIELDS` rows; the
  smart-search constants are sourced from resolved `Settings` (env **and**
  override-file both apply). Restart-required (import-time constants — the
  consuming `SmartSearchTools` is a startup singleton).
- `HAMCP_DASHBOARD_SCREENSHOT_ENGINE_URL` — a `Settings` field that was
  docker/.env-settable but had no panel home (env-only allowlist). Now an
  editable Advanced row; resolved live per capture, so it takes effect
  **without** a restart.

**Audit result (issue ask #2):** those four were the only user-tunable env vars
not already reachable from a UI surface. Everything else read from the
environment is bind / secret / bootstrap / path (`MCP_*`, `FASTMCP_*`,
`SUPERVISOR_*`, `HAMCP_ENV_FILE`, `HA_MCP_CONFIG_DIR`, `XDG_DATA_HOME`,
`HA_MCP_BUILD_VERSION`, `HA_MCP_DISABLE_SETTINGS_UI`) — none are add-on-tunable
— and `DISABLED_TOOLS`/`PINNED_TOOLS` are managed in the Tool Visibility tab.

**Guards against drift (issue ask #3):**

- `test_no_unregistered_direct_env_reads` AST-scans `src/` for direct
  `os.environ`/`os.getenv`/`from os import …` literal reads and fails unless
  each is a registered `Settings` alias or an explicitly documented `ENV_ONLY`
  var (the codified audit of env-only-by-design vars).
- `test_every_advanced_field_has_a_settings_js_label` fails if a new
  `ADVANCED_SETTINGS_FIELDS` row lacks a `settings.js` label (the data-driven
  UI would otherwise render a raw field name with no restart hint).

Also: bounds (`1.0–600.0`, plus `nan`/`inf` rejection) are now enforced on the
env-var path for the budgets — matching the override-file/UI path — so a `≤ 0`
budget can't silently disable the per-id config scan. `partial_reason` hints
and `.env.example` now point at the UI.

## Type of change
- [ ] 🐛 Bug fix
- [x] ✨ New feature
- [ ] 📚 Documentation
- [ ] 🔧 Maintenance/refactor
- [ ] 🧪 Tests only
- [ ] 💥 Breaking change

## Testing
- [ ] I have tested these changes with a LLM agent
- [x] All automated tests pass (`uv run pytest`)
- [x] Code follows style guidelines (`uv run ruff check`)

New unit tests in `tests/src/unit/test_search_time_budget_settings.py` (field
registration/defaults, editable-float in `search` + bounds, env → resolved
value, override-file application, empty/invalid/out-of-range/non-finite
fallback, in-range boundary values, constants track field defaults) and
additions to `test_advanced_settings_coverage.py` (the two guards above + the
scanner self-test + screenshot-URL exposure). Relevant unit suites pass
locally; full e2e runs on CI.

## Checklist
- [x] I have updated documentation if needed

### CODE REVIEW

File: src/ha_mcp/config.py [hunk 1]
```diff
@@ -65,6 +65,24 @@ class Settings(BaseSettings):
...
+    # restart-required behavior the Advanced panel already advertises).
+    automation_config_time_budget: float = Field(
+        30.0, alias="HAMCP_AUTOMATION_CONFIG_TIME_BUDGET"
+    )
+    script_config_time_budget: float = Field(
+        20.0, alias="HAMCP_SCRIPT_CONFIG_TIME_BUDGET"
+    )
+    scene_config_time_budget: float = Field(
+        20.0, alias="HAMCP_SCENE_CONFIG_TIME_BUDGET"
+    )
```
[the remark this answers is not in the document]
Comment (kingpanther13) [reply src/ha_mcp/config.py:84]: Partly taken. I added range enforcement to `_lenient_time_budget` in a261b886 — but as a lenient fall-back rather than a hard raise. An out-of-range value (and empty / unparseable / non-finite) now falls back to the field default with a `logger.warning`, applying the same `_ADVANCED_SETTINGS_BOUNDS` (1.0–600.0) the override-file / UI-POST path uses. NaN and inf are rejected by the same `lo <= v <= hi` test.

I kept it non-raising deliberately: a bad env var shouldn't crash server startup (that preserves the contract of the `_env_float` helper this replaced), and it stays consistent with the other advanced numeric fields — `timeout`, `max_retries`, `entity_search_limit` carry no hard `Field` constraint and enforce range only via `_ADVANCED_SETTINGS_BOUNDS`. Net effect matches your intent (out-of-range rejected at construction) without the startup-crash failure mode.

File: src/ha_mcp/tools/smart_search/_config.py [hunk 2]
```diff
@@ -20,22 +16,17 @@
...
+# the resolved Settings (issue #1538) so the env var, the web Settings UI
+# override file, and the field defaults all flow through one precedence path
+# — and so add-on users (who cannot set raw env vars) can tune them from the
+# Advanced panel. Read once at import as module-level constants; a change
+# takes effect on the next MCP-host restart (the Advanced panel advertises
+# this restart-required behavior).
+_settings = get_global_settings()
+AUTOMATION_CONFIG_TIME_BUDGET = _settings.automation_config_time_budget
+SCRIPT_CONFIG_TIME_BUDGET = _settings.script_config_time_budget
+SCENE_CONFIG_TIME_BUDGET = _settings.scene_config_time_budget
```
[the remark this answers is not in the document]
Comment (kingpanther13) [reply src/ha_mcp/tools/smart_search/_config.py:29]: 📝 Not switching to runtime `self.settings` access — it wouldn't be any more dynamic here.

The consuming `SmartSearchTools` is a **startup singleton**: `server.py:148` builds it once via `create_smart_search_tools(self.client)`, and the `ha_search` / `ha_deep_search` closures capture that instance at registration. Its `self.settings` is therefore itself frozen at startup, so reading `self.settings.<budget>` per call would still not pick up a web-UI change without a restart — identical restart-required semantics to the import-time constant, just less direct.

Issue #1538 explicitly accepts the restart-required behavior for these (they're documented as import-time module constants), and the Advanced panel already shows the "restart required" notice. The constant is sourced from the resolved `Settings` (`get_global_settings()`), so the env var **and** the standalone override file both flow through — it's the simpler equivalent.

File: tests/src/unit/test_advanced_settings_coverage.py [hunk 3]
```diff
@@ -170,6 +174,125 @@ def test_validate_registries_rejects_beta_field_not_in_feature_flags(
...
+    def _is_os_attr(node: ast.expr, attr: str) -> bool:
+        return (
+            isinstance(node, ast.Attribute)
+            and node.attr == attr
+            and isinstance(node.value, ast.Name)
+            and node.value.id == "os"
+        )
+
+    def _is_os_environ(node: ast.expr) -> bool:
+        return _is_os_attr(node, "environ")
```
[the remark this answers is not in the document]
Comment (kingpanther13) [reply tests/src/unit/test_advanced_settings_coverage.py]: ✅ Fixed in 3c1973fe and extended in a261b886. The scanner now resolves per-file import aliases and detects the `from os import environ, getenv` forms (including `as` aliases) — plus `os.environ.pop(...)` — in addition to the `os.environ` / `os.getenv` attribute forms. `test_scanner_detects_all_direct_read_forms` exercises every read form and confirms non-literal (registry-driven) reads are still skipped. (An indirection binding like `e = os.environ; e["X"]` remains out of scope for a static literal scanner; no such pattern exists in `src/` today.)

File: tests/src/unit/test_search_time_budget_settings.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    _reset_global_settings()
+    try:
+        settings = get_global_settings()
+        assert (
+            settings.automation_config_time_budget
+        ) == _config.AUTOMATION_CONFIG_TIME_BUDGET
+        assert settings.script_config_time_budget == _config.SCRIPT_CONFIG_TIME_BUDGET
+        assert settings.scene_config_time_budget == _config.SCENE_CONFIG_TIME_BUDGET
+    finally:
+        _reset_global_settings()
```
[the remark this answers is not in the document]
Comment (kingpanther13) [reply tests/src/unit/test_search_time_budget_settings.py]: ✅ Addressed in 3c1973fe. The test now asserts the constants against the static `Settings` field defaults instead of a live `get_global_settings()` value, so it's deterministic regardless of settings-cache/import state across the session. It still catches a constant that drifts from the Settings default (e.g. re-hardcoding), which is the contract worth locking given the constants are intentionally import-time (see the `_config.py` reply re: the startup singleton).

**Review by kingpanther13 — COMMENTED** [review]
Addressed the review in commits `3c1973fe` + `a261b886`:

- Extended the env-read guard to catch `from os import …` / `as` aliases / `os.environ.pop()` (#3), with a `test_scanner_detects_all_direct_read_forms` self-test.
- Made the constants test deterministic against the `Settings` field defaults rather than a live `get_global_settings()` value (#4).
- Added lenient `_ADVANCED_SETTINGS_BOUNDS` enforcement on the env-var path so it matches the override-file / UI path — fixing a silent scan-disable when `budget <= 0` (and rejecting `nan`/`inf`). This is the spirit of #1, kept non-raising so a bad env var doesn't crash startup (consistent with `timeout` / `entity_search_limit`).
- Kept the import-time constant rather than per-call `self.settings` (#2): `SmartSearchTools` is a startup singleton (`server.py:148`), so reading `self.settings` per call would be no more dynamic — same restart-required semantics, which #1538 explicitly accepts.

CI is green.

**Review by Patch76 — APPROVED** [review]
Approving. The env-var audit and the AST-based CI guard are the right shape for #1538's three asks — the scanner self-test plus the ENV_ONLY dead-entry check make it hard to regress, and CI is green across e2e/docker/unit.

One non-blocking item for your call:

### [nit] `dashboard_screenshot_engine_url` skips its scheme validator on the new UI/override path

**What:** `validate_dashboard_screenshot_engine_url` enforces `http(s)://`, but only at model construction (the env-var path). Adding the field to `ADVANCED_SETTINGS_FIELDS` routes UI/override-file writes through `_save_advanced_settings` (settings_ui.py) and `_apply_advanced_overrides` (config.py), which validate str fields by type + null-byte + bounds + choices only. This field has no bounds/choices entry, and `setattr` doesn't re-validate (`validate_assignment` is unset).

**Effect:**

- env path: `ftp://x` or a bare hostname → rejected loudly at startup
- UI/override path (new here): same value → accepted silently, applied live on next start

That reintroduces the 0-byte capture-fail the validator exists to prevent, and the UI path is the only one add-on users have for this field. The three time-budget floats don't have this gap — their range is a `_ADVANCED_SETTINGS_BOUNDS` entry both write paths check.

**Fix (localized):** replicate the `http(s)://` check in `_save_advanced_settings` + `_apply_advanced_overrides`, or give the field a validation hook the override path already honors. It's the only editable str advanced field whose validator isn't mirrored by a bounds/choices entry (`backup_hint`/`log_level` are choices-covered).

Fine to defer.

**Comment by kingpanther13** [comment]
I"m not too concerned about that, it's set automatically in the addon version anyways and it's labeled as an advanced setting, so theoretically nobody should be messing with it anyways unless they really know what they're doing
